From b1671134f1e09ac030dba4e6ae1f667f39c9e6e2 Mon Sep 17 00:00:00 2001
From: MeiChia Chiu <meichia.chiu@mediatek.com>
Date: Thu, 26 Oct 2023 10:08:10 +0800
Subject: [PATCH 032/146] mtk: mt76: mt7996: Fixed null pointer dereference
 issue

mtk: mt76: add sanity check to prevent kernel crash

wcid may not be initialized when mac80211 calls mt76.tx and it would lead to
kernel crash.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/main.c |  7 +++++++
 tx.c          | 10 +++++++++-
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 7bdccdf7..4dbda52f 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1057,9 +1057,16 @@ static void mt7996_link_sta_rc_update(struct ieee80211_hw *hw,
 				      u32 changed)
 {
 	struct ieee80211_sta *sta = link_sta->sta;
+	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
 	struct mt7996_phy *phy = mt7996_hw_phy(hw);
 	struct mt7996_dev *dev = phy->dev;
 
+	if (!msta->vif) {
+		dev_warn(dev->mt76.dev, "Un-initialized STA %pM wcid %d in rc_work\n",
+			 sta->addr, msta->wcid.idx);
+		return;
+	}
+
 	mt7996_sta_rc_work(&changed, sta);
 	ieee80211_queue_work(hw, &dev->rc_work);
 }
diff --git a/tx.c b/tx.c
index ce193e62..c24729d6 100644
--- a/tx.c
+++ b/tx.c
@@ -352,8 +352,16 @@ mt76_tx(struct mt76_phy *phy, struct ieee80211_sta *sta,
 	if ((info->flags & IEEE80211_TX_CTL_TX_OFFCHAN) ||
 	    (info->control.flags & IEEE80211_TX_CTRL_DONT_USE_RATE_MASK))
 		head = &wcid->tx_offchannel;
-	else
+	else {
+		if (!wcid->tx_pending.prev || !wcid->tx_pending.next) {
+			dev_warn(phy->dev->dev, "Un-initialized STA %pM wcid %d in mt76_tx\n",
+				 sta->addr, wcid->idx);
+
+			ieee80211_free_txskb(phy->hw, skb);
+			return;
+		}
 		head = &wcid->tx_pending;
+	}
 
 	spin_lock_bh(&head->lock);
 	__skb_queue_tail(head, skb);
-- 
2.45.2

